Sunlight angle

jbytheway opened the pull request:
#### Summary
Features "Calculate angle of sunlight as a function of time"

#### Purpose of change
Calculate sunlight angle using more realistic formulae.

This allows for more realistic sunrise and sunset times.  It also allows for many other changes like calculating solar incidence on panels, directional shadows, etc.

#### Describe the solution
Use 'real' solar position formulae, adapted to our situation.
* Tweak them to cope properly with the variable season length.
* Ignore precession, but do take into account the mean anomaly, so that the path of noon throughout the year forms an analemma shape.

As currently written, this only includes the code to perform the calculations, and lots of unit tests.  No impact on gameplay.

#### Describe alternatives you've considered
More or less complex calculations are possible.

#### Testing
Lots of unit tests included.

#### Additional context
I wrote this months ago.  I was holding back on PRing it until the current feature freeze was over, but since (to my surprise) #47570 got published with very similar code, I felt I should publish this so we can work to combine the best of both approaches.

BrettDong commented on src/line.h:
```diff
@@ -11,6 +11,8 @@
 #include "point.h"
 #include "units_fwd.h"
 
+class rl_vec2d;
```
```suggestion
struct rl_vec2d;
```

actual-nh commented:
So this is taking into account the elliptical orbit of the earth around the sun, and the axial tilt, with adjustments for a 364-day year (with 91-day seasons)?

Hirmuolio commented:
Your math is definitely more accurate than mine.
I made some rough simplifications. Orbits are perfect circles, Earth is perfect sphere, sun is always directly sout at noon, local time is solar time and so on. No analemma shape with my way.

I know how  inacurate the azimuth angle with my math is but when I compared my altitudes to this tool they were within few degrees which I think is "probably close enough" https://www.esrl.noaa.gov/gmd/grad/solcalc/azel.html.

But the approximations make calculating times for arbitrary angles much simpler. Like the time of next dawn and whether dawn even happens today.
Maybe my math could be used for those situations and your math for theactual sun angles. 

I'll try to smash them both in there at the same time to compare how much their results differ.

actual-nh commented:
Clang 12 (OSX 10.15, Tiles):
```
Error: ../tests/vehicle_part_test.cpp:34:19: error: unused variable 'midnight' [-Werror,-Wunused-variable]
static time_point midnight = calendar::turn_zero + 0_hours;
```
Umm... vehicle part test? Solar panels or something?

Fris0uman commented:
Could it be jsonised a bit? Changing sun path would be usefull for overhauld mods to put the game either somewhere else on earth or on another planet like for Aftershock

Hirmuolio commented:
I managed to smash them together but I get some odd results from your equation.

Year 1 day 1 at 11:00
Coordinates: 40°N 0°E

My math: Altitude 25.5° azimuth 15.3°
Your math: Altitude 48.5° azimuth 335°

[NOAA calc](https://www.esrl.noaa.gov/gmd/grad/solcalc/azel.html) for jan 1 2021 11:00: Altitude 25.34° azimuth 163.78°

All three use different definition for the azimuth :s. But the altitudes should be directly comparable.
Maybe I messed something when copy pasting your code. Could you check what you get for this?

jbytheway commented:
Year 1 day 1 is not January 1st, it's the Spring equinox (I think that's what I picked, anyway).  I'm guessing that's where the difference arises?  It makes sense that the altitude is higher on the equinox than on Jan 1st.

jbytheway commented:
@Fris0uman Yes, it could be jsonised, but I thought it was best to keep it simple and hardcoded for the first implementation.  One of my very-long-term ideas is support for teleporting to other planets/dimensions, and each would need to define its own astronomical stuff.  We could have multiple moons; maybe even multiple stars.  Lots of blue-sky ideas, but trying to keep the initial PR to a minimum viable implementation :).

jbytheway commented:
> But the approximations make calculating times for arbitrary angles much simpler. Like the time of next dawn and whether dawn even happens today.
> Maybe my math could be used for those situations and your math for theactual sun angles.

Yeah, I hadn't decided exactly what to do about the calculations for dawn and dusk.  My guess was that inverting the calculation was impossible, so I was planning to do a binary search on time to find the point where the appropriate event happened.  Then we can cache the last few times (maybe just the last one time) calculated so we don't need to repeat the potentially expensive search.  I imagine that when the game asks for sunrise time it's almost always asking for the same one it asked for last time, so it should be easy to make a useful cache.

jbytheway commented:
> So this is taking into account the elliptical orbit of the earth around the sun, and the axial tilt, with adjustments for a 364-day year (with 91-day seasons)?

Yes, except its adapted for the actual year length, for whatever season the player chooses.  I hadn't decided how to handle the eternal season option, but it shouldn't actually be too hard, I think.

> EDIT: To what degree does the current code "pay attention to" the elliptical orbit? Just for how the sun moves, or for how much light it gives?

This code only calculates the direction to the Sun, but if we wanted the distance to the Sun too, then it can be found as a function of the mean anomaly (which is already used here) via one simple additional calculation, as seen on https://en.wikipedia.org/wiki/Position_of_the_Sun.

jbytheway commented:
> ```
> Error: ../tests/vehicle_part_test.cpp:34:19: error: unused variable 'midnight' [-Werror,-Wunused-variable]
> static time_point midnight = calendar::turn_zero + 0_hours;
> ```
> 
> Umm... vehicle part test? Solar panels or something?

I think that failure is unrelated to this code; I didn't touch that file.

jbytheway commented:
> I managed to smash them together but I get some odd results from your equation.
> 
> Year 1 day 1 at 11:00
> Coordinates: 40°N 0°E
> 
> My math: Altitude 25.5° azimuth 164.7°
> Your math: Altitude 48.5° azimuth 335°
> 
> [NOAA calc](https://www.esrl.noaa.gov/gmd/grad/solcalc/azel.html) for jan 1 2021 11:00: Altitude 25.34° azimuth 163.78°
> 
> Different definition for the azimuth :s (?). But the altitudes should be directly comparable.
> Maybe I messed something when copy pasting your code. Could you check what you get for this?

Using the same NOAA calc website for March 20th, 2000, I get

Altitude 47.22, azimuth 154.75 = 334.75-180

so that seems to match my calculations fairly well.

I wasn't sure what definition to use for Azimuth; I found conflicting sources.  Happy to change it to be zero-at-north if appropriate.

Hirmuolio commented:
Ok I have adjusted the code to assume day 1 = march 20 (79th day of year). The two now give nearly identical results.

Some comparisons at 10:00

| Day of year  | Altitude (me) | Altitude (you) | Azimuth (me) | Azimuth (you) |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| 79  | 40.9 | 42.0 | 139.0 | 135.6  | 
| 169  | 60.0 | 59.7 | 114.6 | 114.0  | 
| 220  | 54.8 | 53.3 | 124.3 | 123.6  | 
| 311  | 26.6 | 27.9 | 148.0 | 152.0  | 
| 37  | 27.4 | 27.3 | 147.6 | 143.3  | 

The differences are so small that player is unlikely to notice them so the two different styles can be mixed. My equations for reverse calculations (when is the next dawn etc.) and your math for the current exact position of the Sun.

jbytheway commented:
There is a Wikipedia page for exactly calculating sunrise (https://en.wikipedia.org/wiki/Sunrise_equation) so I think it *should* be possible to do it 'properly' without resorting to binary search.  I might try implementing it, but not sure when I'll get round to it.  Having an approximation based on your code for the initial version is probably fine.

actual-nh commented:
> > So this is taking into account the elliptical orbit of the earth around the sun, and the axial tilt, with adjustments for a 364-day year (with 91-day seasons)?
> 
> Yes, except its adapted for the actual year length, for whatever season the player chooses. I hadn't decided how to handle the eternal season option, but it shouldn't actually be too hard, I think.

Treat it as no axial tilt and a circular orbit (taking 364 days, I would guess)?

> > EDIT: To what degree does the current code "pay attention to" the elliptical orbit? Just for how the sun moves, or for how much light it gives?
> 
> This code only calculates the direction to the Sun, but if we wanted the distance to the Sun too, then it can be found as a function of the mean anomaly (which is already used here) via one simple additional calculation, as seen on https://en.wikipedia.org/wiki/Position_of_the_Sun.

Excellent!

Hirmuolio commented:
That dawn equation is exactly what I use in `time_point sun_at_angle()`. I just use approximation for the declination of the Sun and hour angle in it.

Perhaps your `sun_azimuth_altitude` function could be split to parts: Declination+hour angle in first part and altitude +azimuth. This way my equations would be using the same values as your equations.

jbytheway commented:
> > Yes, except its adapted for the actual year length, for whatever season the player chooses. I hadn't decided how to handle the eternal season option, but it shouldn't actually be too hard, I think.
> 
> Treat it as no axial tilt and a circular orbit (taking 364 days, I would guess)?

No, that would only work for perpetual Spring or Autumn, not Summer or Winter.  We'd need to separate out the inputs related to the time of year and the time of day, and fix the time of year input.  Except that's not quite right, because then there'd be one extra day per year (or maybe one fewer).  Maybe easier would be to just always shift the time_point used for the calculation to a specific day of the year.  That would lead to a dicontinuity of the Sun's position at midnight, but I guess no one would notice :).

jbytheway commented:
> That dawn equation is exactly what I use in `time_point sun_at_angle()`. I just use approximation for the declination of the Sun and hour angle in it.
> 
> Perhaps your `sun_azimuth_altitude` function could be split to parts: Declination+hour angle in first function and altitude +azimuth on second function. This way my equations would be using the same values as your equations.

Yep, I like that plan.

Hirmuolio commented:
Vhecked my equations for what is needed where.

Sun declination is needed for:
* When is Sun at altitude X
* Current Sun altitude
* Does Sun reach altitude X today

Sun hour angle is only needed in the Sun altitude calculation. 

So the declination would be best on its own and hour angle can stay in the altitude/azimuth angle function.

actual-nh commented:
> > > Yes, except its adapted for the actual year length, for whatever season the player chooses. I hadn't decided how to handle the eternal season option, but it shouldn't actually be too hard, I think.
> > 
> > Treat it as no axial tilt and a circular orbit (taking 364 days, I would guess)?
> 
> No, that would only work for perpetual Spring or Autumn, not Summer or Winter. We'd need to separate out the inputs related to the time of year and the time of day, and fix the time of year input. Except that's not quite right, because then there'd be one extra day per year (or maybe one fewer). Maybe easier would be to just always shift the time_point used for the calculation to a specific day of the year. That would lead to a dicontinuity of the Sun's position at midnight, but I guess no one would notice :).

One could use a date input that went from the start of a given season to the end of it, then went backward (while moving time/the conceptual spin of the Earth forward, of course). If seasons had some progression of temperature, etc conditions within the season, that would go along with an eternal season swinging from one end of the season to the other and back again.

Hirmuolio commented:
For eternal season I believe taking the time (hour:minute:second) normally and date from `calendar::start_of_game` should work.

Fancier method would be to use modulus to limit the date within the same season as `calendar::start_of_game`. This way the sun motion would loop with the motion of that season.

jbytheway commented:
> Vhecked my equations for what is needed where.
> 
> Sun declination is needed for:
> 
>     * When is Sun at altitude X
> 
>     * Current Sun altitude
> 
>     * Does Sun reach altitude X today
> 
> 
> Sun hour angle is only needed in the Sun altitude calculation.
> 
> So the declination would be best on its own and hour angle can stay in the altitude/azimuth angle function.

The new function should probably return right ascension and declination, since they're basically computed together (and right ascension is needed for hour angle).

actual-nh commented:
> For eternal season I believe taking the time (hour:minute:second) normally and date from `calendar::start_of_game` should work.

Would it be always that season?

> Fancier method would be to use modulus to limit the date within the same season as `calendar::start_of_game`. This way the sun motion would loop with the motion of that season.

Modulus has the problem of a very large jump at the end of the season - why I suggested going backward in terms of the solar movement not from Earth's spin.

jbytheway commented:
I've rebased this PR on 0.F-dev, and added some of the thing's from @Hirmuolio's PR #47570.

Specifically, sunrise and sunset are now based on the new calculations.  I did a slightly more complex version of the calculation than in the other PR, in order to get the times computed for these events to actually match the times when the Sun is at the horizon.

Apart from that, most of the code is lifted from the other PR.

The issue of how to handle eternal seasons remains unresolved.

jbytheway commented:
I checked and the old code actually used the modulus approach to day lengths, and no one seems to have complained about the discontinuity.  Of course, I think the eternal season option is rarely used.

jbytheway commented on src/calendar.cpp:
```diff
@@ -101,149 +93,308 @@ moon_phase get_moon_phase( const time_point &p )
+    units::angle azimuth, altitude;
+    std::tie( azimuth, altitude ) = sun_azimuth_altitude( t, location, timezone_boston );
+    if( altitude <= 0_degrees ) {
+        return cata::nullopt;
```
Yes.  That's explained in the comment on the declaration in `calendar.h`.  I'm never sure whether to put such things in the header or cpp file, and I don't like to repeat it...  But I can add something brief here.

Brambor replied:
> Yes. That's explained in the comment on the declaration in `calendar.h`. I'm never sure whether to put such things in the header or cpp file, and I don't like to repeat it... But I can add something brief here.

My teacher says: To .h file put comments regarding usage. To .cpp put comments regarding implementation.

In here, if you want to, the comment can be even more explanatory:
```
// Sun below horizon (nighttime)
```

jbytheway commented on src/calendar.cpp:
```diff
@@ -101,149 +93,308 @@ moon_phase get_moon_phase( const time_point &p )
-    static const std::array<int, 4> start_hours = { { sunset_equinox, sunset_summer, sunset_equinox, sunset_winter, } };
-    const size_t season = static_cast<size_t>( season_of_year( p ) );
-    cata_assert( season < start_hours.size() );
+    const rl_vec3d intermediate(
```
It's something between equatorial and horizontal coordinates.  If it has a name, I don't know what it is.  I'll add a link to a reference.

jbytheway replied:
...and I've renamed `hor` to `horizontal`.

jbytheway commented:
@actual-nh thanks for the review.  I think I've addressed all the things you raised / suggested.

jbytheway commented:
I've taken this out of draft now.  I don't know of any remaining problems, although it could certainly stand to have more testing.

For the eternal season I ended up going with @Hirmuolio's first suggestion.  It just repeats the day of `calendar::start_of_game` forever.  So there is a small discontinuity at midnight, but it shouldn't be obvious.

Things that might still be worth adding (both of which already exist in #47570), but I don't think are critical:
* Tests for calculations at other latitudes.
* A sextant or some other way for the player to directly observe the Sun's position.

actual-nh commented:
A couple of things, one just for amusement:
- Sidereal time is generally defined as relative to "fixed stars" - amusing in the context of #48244...
- Re the clang-tidy build error, I note that it seems to be using a python 2.7 library directory; perhaps this can be overridden?

jbytheway commented:
The clang-tidy error would be fixed by #48739.

jbytheway commented:
OK, now that the CI is in better shape I was finally able to fix the clang-tidy errors on this branch.  I think it's now in a reasonable state to merge.

Hirmuolio commented:
I think it should be possible to change the latitude. Probably best to have it as an external option so mods can change it. Or  it could be option in the world creation.

Longitude is a bit reduntant when used with the timezone as they should almost completely cancel each other out. So I think it would be best to just disregard it and do the math with assumption that player is at 0° E and timezone is UTC+0.

It also seems like you took some older code for the sunlight amount.
I changed the "Linear increase from 0° to 70° degrees light increases from 60 to 125 brightness." into "Linear increase from 0° to 30° degrees light increases from 70 to 125 brightness."
This is a bit iffy since vision and solar panels use the same sunlight. But for vision purposes sunlight reaches maximum value at pretty low angle but for solar panels the maximum should be at around 90°.

jbytheway commented:
> I think it should be possible to change the latitude. Probably best to have it as an external option so mods can change it. Or it could be option in the world creation.

Yes, I agree, but not in the initial implementation.  For one thing, I think the sunrise/sunset calculations probably break down near the poles, so I don't want to offer that freedom until it's more carefully tested.  And I want to make this a minimal change focusing on just the maths to simplify review; adding configuration would be more diff.

> Longitude is a bit reduntant when used with the timezone as they should almost completely cancel each other out. So I think it would be best to just disregard it and do the math with assumption that player is at 0° E and timezone is UTC+0.

If we ever want to support e.g. teleporting to another place on Earth it would be a lot of work to retrofit this feature, so I'd rather keep it now it's working.  On the other hand, I can see your point.

> Also the way timezone and longitude/latitude enter the calculations as function arguments is a bit confusing since the functions aren't called with them. They just pop in.

Agreed.  I think this would become cleaner if the parameters were pulled out to JSON, but as stated I went with this interim solution for now to make this a minimum viable change.

> It also seems like you took some older code for the sunlight amount.
> I changed the "Linear increase from 0° to 70° degrees light increases from 60 to 125 brightness." into "Linear increase from 0° to 30° degrees light increases from 70 to 125 brightness."
> This is a bit iffy since vision and solar panels use the same sunlight. But for vision purposes sunlight reaches maximum value at pretty low angle but for solar panels the maximum should be at around 90°.

I didn't take older code; I selected new values to (very roughly) fit the previously established brightnesses at solstice and equinoxes of 75 / 100 / 125.  I believe those numbers were chosen for some reason at some point (although I didn't really research the history).  Mostly I'm trying not to change too many different things at once by sticking to that existing convention.  The brightness values are very much arbitrary, though, as you've pointed out it's not really clear what their real-world meaning is because they are used for various purposes.  So I have no problem changing them, but I suspect it's mostly an opportunity for bike-shedding.

Hirmuolio commented:
Lets keep the longitude then for potential future use. But timezone is still unnecessary to have as a separate thing. Instead we could just use "local time timezone" instead of "proper timezone".
timezone = longitude / 360° * 24h (west longitude is negative, east is positive).

The problem with the current linear 0° to 70° for 60 to 125 is that in northern location during winter it makes even middle of the day noticeably dark. The problem isn't too noticeable at the current Boston latitude and since the latitude can't be changed yet that can be fixed later.

Also it seems like the game recalculates the sun angle constantly (for drawing the world?). Not sure how much processing it takes since the CPU profiler is till not working on my system (Microsoft fix it already).
The angle should probably be cached so it isn't constantly recalculated. And since the Sun moves so slow it could be cached for a longer time too so it isn't recalculated every turn. At dusk it seems that the brighness changes noticeably once every few tens of seconds so recalculating Sun postion once every 7 turns or so should be unnoticeable to player.

jbytheway commented:
> Lets keep the longitude then for potential future use. But timezone is still unnecessary to have as a separate thing. Instead we could just use "local time timezone" instead of "proper timezone".
> timezone = longitude / 360° * 24h (west longitude is negative, east is positive).

That's probably a reasonable simplification for now.  I'll give it a go and see how it turns out.

> The problem with the current linear 0° to 70° for 60 to 125 is that in northern location during winter it makes even middle of the day noticeably dark. The problem isn't too noticeable at the current Boston latitude and since the latitude can't be changed yet that can be fixed later.

Agreed; this probably needs further balance work, but best in a later PR.

> Also it seems like the game recalculates the sun angle constantly (for drawing the world?). Not sure how much processing it takes since the CPU profiler is till not working on my system (Microsoft fix it already).
> The angle should probably be cached so it isn't constantly recalculated. And since the Sun moves so slow it could be cached for a longer time too so it isn't recalculated every turn. At dusk it seems that the brighness changes noticeably once every few tens of seconds so recalculating Sun position once every 7 turns or so should be unnoticeable to player.

It's just a few trig function calls.  It should be negligible relative to things like drawing the screen.  I doubt optimizing it is worth the trouble.  But I'll do some profiling to verify that.

(In related news, I have been working on a follow-up PR implementing shadows based on the angle of sunlight; that's why I've been a bit quiet on this one.  I have something that works as a crude first draft, but is much too slow.  Performance there will be a genuine concern.  But that will also have to wait for a future PR)

jbytheway commented:
OK, I've rebased on latest 0.F and removed timezone from most of the APIs (including all the `calendar.h` functions).  It simplifies one function.  I could also simplify `sidereal_time_at` but it would require rethinking the function a bit more and I don't think it's necessary.

The biggest disadvantage of this change, IMO, is that the test cases for sunrise and sunset have moved further from the 'true' Boston sunrise and sunset times I was quoting for comparison's sake.  But that's not really a big deal.  I could find some random town with a longitude that falls exactly on the -5 hours meridian and use that instead of Boston for the location, but that's probably getting a little silly.

Still need to do some performance tests.

jbytheway commented:
After some investigation, light levels are already cached.  That cache is cleared in `cata_tiles::draw`, so probably more often than necessary, but only once per screen draw, which is rarely enough that it shouldn't be a problem.

I profiled a release build and a debug build.  In both cases `game::natural_light_level`, `sun_azimuth_angle`, and `sun_ra_declination` were each at most 0.01% of recorded samples.

So I think the performance is not an issue.

jbytheway commented:
Another test, this time waiting to pass time as fast as possible rather than moving, which I thought might show more demand on thes functions.  Indeed it did.  This time `game::natural_light_level` was 0.17%, and `sun_azimuth_altitude` was 0.31%.  So, more significant, but IMO still not a concern.

jbytheway commented:
CI failures are unrelated; fixed in #49351.

scorpion451 commented:
Speaking with my artist hat on as a "future development from here" thing, when calculating sunset/sunrise light levels and deep outdoor shadows you'll want to build in a minumum light ambient light level to represent atmospheric diffusion and scattering. 


Basically there's a non-linear falloff due to light bouncing around in the atmosphere, so shadows at around noontime can be as bright as late-evening direct lighting on a clear day, and it doesn't actually get dark until the sun is about 18* below the horizon, which can take an hour or two.
